File: Java/Roman to Integer.java
// 3ms, 100.00%; 41.1MB, 5.48%
class Solution {
    static public int romanToInt(String s) {
        int res = 0, last = 0;
        for (int i = s.length() - 1; i > -1; --i) {
            int tmp = getValue(s.charAt(i));
            // any reverse order, like 'IV', means `valueOf('V') - valueOf('I')`
            res += tmp * (tmp >= last ? 1 : -1);
            last = tmp;
        }
        return res;
    }

    final static private String order = "IVXLCDM";
    final static private int[] vals = { 1, 5, 10, 50, 100, 500, 1000 };

    static private int getValue(char c) {
        return vals[order.indexOf(c)];
    }
}

File: Go/Roman to Integer.go
// 8ms, 59.35%; 3.1MB, 14.29%
var hashTable = map[byte][2]int{'I': {1, 1}, 'V': {2, 5}, 'X': {3, 10},
	'L': {4, 50}, 'C': {5, 100}, 'D': {6, 500}, 'M': {7, 1000}}

func romanToInt(s string) int {
	res, last := 0, 0
	for i := len(s) - 1; i > -1; i-- {
		vals, _ := hashTable[s[i]]
		if vals[0] < last {
			res -= vals[1]
		} else {
			res += vals[1]
		}
		last = vals[0]
	}
	return res
}
